FantasmoSDK/Classes/Filters/FMInputQualityFilter.swift

```diff
@@ -0,0 +1,107 @@
+//
+//  FMInputQualityFilter.swift
+//  FantasmoSDK
+//
+//  Created by lucas kuzma on 3/30/21.
+//
+
+import ARKit
+
+enum FMFilterRejectionReason {
+    case pitchTooLow
+    case pitchTooHigh
+    case movingTooFast
+    case movingTooLittle
+}
+
+enum FMFilterResult: Equatable {
+    case accepted
+    case rejected(reason: FMFilterRejectionReason)
+}
+
+protocol FMFrameFilter {
+    func accepts(_ frame: ARFrame) -> FMFilterResult
+}
+
+// MARK:-
+
+public class FMInputQualityFilter {
+    
+    var delegate: FMLocationDelegate?
+    
+    /// the last time a frame was accepted
+    var lastAcceptTime = clock()
+    /// number of seconds after which we force acceptance
+    var acceptanceThreshold = 6.0
+    
+    /// the last time we issued a behavior request
+    var lastRequestTime = clock()
+    /// minimum number of seconds that must elapse between behavior requests
+    var throttleThreshold = 2.0
+    /// the number of times a rejection occurs that should prompt a behavior request
+    var incidenceThreshold = 30
+    
+    /// filter collection, in order of increasing computational cost
+    let filters: [FMFrameFilter] = [
+        FMAngleFilter(),
+        FMMovementFilter(),
+        FMBlurFilter(),
+    ]
+    
+    var rejections: [FMFilterRejectionReason: Int] = [:]
+
+    func startFiltering() {
+        resetAcceptanceClock()
+    }
+
+    func resetAcceptanceClock() {
+        lastAcceptTime = clock()
+    }
+    
+    /// run ARFrame through quality filter collection
+    func accepts(_ frame: ARFrame) -> Bool {
+
+        if !shouldForceAccept() {
+            // run frame through filters
+            for filter in filters {
+                if case let .rejected(rejection) = filter.accepts(frame) {
+                    addRejection(rejection)
+                    notifyIfNeeded(rejection)
+                    return false
+                }
```

Comment:
Our plain (non compound) filters (filter rules) are all "sequence" filters, that is each filter supposes that it will receive all frames from sequence. This is important for proper work of "sequence filter". So, even if some plain filter fails we should also pass frame to all other plain filters to ensure that no frame was dropped for some plain filter.

Reply:
In the case of the two stateful filters, we do not care if we capture every frame. Afaict they deliver desired results even when considering only the frames that make it through the previous filters. Your point is a good one, and we should consider it if we add a filter that really does need every frame. The current design is intended to minimize computational cost, particularly that incurred by the blur filter.